File: common/src/main/java/org/valkyrienskies/clockwork/content/contraptions/sequenced_seat/SequencedSeatRenderer.java
package org.valkyrienskies.clockwork.content.contraptions.sequenced_seat;

import com.mojang.blaze3d.vertex.PoseStack;
import com.simibubi.create.AllBlockPartials;
import com.simibubi.create.content.contraptions.base.KineticTileEntity;
import com.simibubi.create.content.contraptions.base.KineticTileEntityRenderer;
import com.simibubi.create.foundation.render.CachedBufferer;
import com.simibubi.create.foundation.render.SuperByteBuffer;
import com.simibubi.create.foundation.utility.AngleHelper;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.core.Direction;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import org.valkyrienskies.clockwork.ClockWorkPartials;

public class SequencedSeatRenderer extends KineticTileEntityRenderer {

    private SequencedSeatBlockEntity te;

    public SequencedSeatRenderer(BlockEntityRendererProvider.Context context) {
        super(context);
    }

    @Override
    protected void renderSafe(KineticTileEntity te, float partialTicks, PoseStack ms, MultiBufferSource buffer,
                              int light, int overlay) {
        super.renderSafe(te, partialTicks, ms, buffer, light, overlay);
        if (!(te instanceof SequencedSeatBlockEntity))
            return;

        this.te = (SequencedSeatBlockEntity) te;
        SequencedSeatBlockEntity seat = this.te;
        final Direction facing = te.getBlockState()
                .getValue(BlockStateProperties.HORIZONTAL_FACING);

        SuperByteBuffer joystick = CachedBufferer.partialFacing(ClockWorkPartials.JOYSTICK, te.getBlockState(), facing.getOpposite());
        SuperByteBuffer buttonone = CachedBufferer.partialFacing(ClockWorkPartials.BUTTON_ONE, te.getBlockState(), facing.getOpposite());
        SuperByteBuffer buttontwo = CachedBufferer.partialFacing(ClockWorkPartials.BUTTON_TWO, te.getBlockState(), facing.getOpposite());

//        superBuffer.rotateCentered(Direction.EAST, AngleHelper.rad(-90 - AngleHelper.verticalAngle(facing)));
//        superBuffer.renderInto(ms, buffer.getBuffer(RenderType.solid()));
        joystick.light(light).renderInto(ms, buffer.getBuffer(RenderType.solid()));
        buttonone.light(light).renderInto(ms, buffer.getBuffer(RenderType.solid()));
        buttontwo.light(light).renderInto(ms, buffer.getBuffer(RenderType.solid()));
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/kinetics/sequenced_seat/SequencedSeatRenderer.kt
package org.valkyrienskies.clockwork.content.kinetics.sequenced_seat

import com.mojang.blaze3d.vertex.PoseStack
import com.simibubi.create.AllPartialModels
import com.simibubi.create.content.kinetics.base.HorizontalKineticBlock
import com.simibubi.create.content.kinetics.base.KineticBlockEntityRenderer
import com.simibubi.create.foundation.render.CachedBufferer
import com.simibubi.create.foundation.render.SuperByteBuffer
import com.simibubi.create.foundation.utility.Iterate
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import org.valkyrienskies.clockwork.ClockworkPartials

class SequencedSeatRenderer(context: BlockEntityRendererProvider.Context) :
    KineticBlockEntityRenderer<SequencedSeatBlockEntity>(context) {
    private var te: SequencedSeatBlockEntity? = null
    override fun renderSafe(
        te: SequencedSeatBlockEntity,
        partialTicks: Float,
        ms: PoseStack,
        buffer: MultiBufferSource,
        light: Int,
        overlay: Int
    ) {
        super.renderSafe(te, partialTicks, ms, buffer, light, overlay)
        this.te = te
        val seat = this.te
        val facing = te.blockState
            .getValue(BlockStateProperties.HORIZONTAL_FACING)

        renderRotatingBuffer(te, getRotatedModel(te, te.blockState), ms,
            buffer.getBuffer(RenderType.solid()), light)

        val joystick = CachedBufferer.partialFacing(ClockworkPartials.JOYSTICK, te.blockState, facing.opposite)
        val buttonone = CachedBufferer.partialFacing(ClockworkPartials.BUTTON_ONE, te.blockState, facing.opposite)
        val buttontwo = CachedBufferer.partialFacing(ClockworkPartials.BUTTON_TWO, te.blockState, facing.opposite)

        // superBuffer.rotateCentered(Direction.EAST, AngleHelper.rad(-90 - AngleHelper.verticalAngle(facing)));
        // superBuffer.renderInto(ms, buffer.getBuffer(RenderType.solid()));
        //joystick.light(light).renderInto(ms, buffer.getBuffer(RenderType.solid()))
        //buttonone.light(light).renderInto(ms, buffer.getBuffer(RenderType.solid()))
        //buttontwo.light(light).renderInto(ms, buffer.getBuffer(RenderType.solid()))

    }

    override fun getRotatedModel(te: SequencedSeatBlockEntity, state: BlockState): SuperByteBuffer {
        return CachedBufferer.partialFacing(
            AllPartialModels.SHAFT_HALF, state, Direction.DOWN
        )
    }
}